File: day3/puzzle2.py
""" Task: Determine which claim has no overlaps """
from collections import namedtuple 

import numpy as np 

Claim = namedtuple("Claim", ["id", "left_offset", "top_offset", "width", "height"])

def check_claim(claim, fabric):
    lo = claim.left_offset 
    to = claim.top_offset
    w = claim.width 
    h = claim.height 
    return ~np.any(fabric[to:to+h, lo:lo+w] > 1)

def find_no_overlap(claims, fabric):
    for claim in claims:
        if check_claim(claim, fabric):
            return claim.id 

def main():
    
    edge = 1000
    fabric = np.zeros((edge, edge), int)
    claims = list()
    with open("input.txt") as f:
        for line in f:
            claim_id = int(line.split("@")[0][1:])
            offsets = line.split("@")[1].split(":")[0].split(",")
            lo, to = [int(off) for off in offsets]
            size = line.split(":")[1].split("x")
            w, h = [int(s) for s in size]
            fabric[to:to+h, lo:lo+w] += 1
            claims.append(Claim(id=claim_id, left_offset=lo, top_offset=to, width=w, height=h))
    
    print(find_no_overlap(claims, fabric))


if __name__ == "__main__":
    main()

File: day3/puzzle2.jl
""" Task: Determine which claim does not overlap with any other """

struct Claim
    id::Int 
    leftoffset::Int
    topoffset::Int
    width::Int
    height::Int 
end

const edge = 1000

function checkclaim(claim, fabric)
    for i in claim.leftoffset+1:claim.leftoffset+claim.width
        for j in claim.topoffset+1:claim.topoffset+claim.height
            if fabric[i, j] > 1
                return false
            end
        end
    end
    return true
end

function findnooverlap(claims, fabric)
    for claim in claims
        thisone = checkclaim(claim, fabric)
        if thisone 
            return claim.id 
        end
    end
    # Return -1 if failure
    return -1::Int 
end

function parseclaim!(line, fabric)
    id = parse(Int, split(line, "@")[1][2:end])
    lo, to = parse.(Int, split(split(split(line, "@")[2], ":")[1], ","))
    w, h = parse.(Int, split(split(line, ":")[2], "x"))
    for i in lo+1:lo+w
        for j in to+1:to+h
            fabric[i, j] += 1
        end
    end
    return Claim(id, lo, to, w, h)
end

function main() 

    # Process the claims
    fabric = zeros(Int, edge, edge)
    claims = open("input.txt", "r") do file 
        claims = Array{Claim, 1}()
        for line in eachline(file)
            push!(claims, parseclaim!(line, fabric))
        end 
        claims 
    end

    println(findnooverlap(claims, fabric))

end

@time main()
